README.org
# SPDX-FileCopyrightText: Idiap Research Institute
# SPDX-FileContributor: Enno Hermann <enno.hermann@idiap.ch>
# SPDX-License-Identifier: GPL-3.0-only

* Idiap Speech Processing Exercises

Set of exercises on automatic speech processing.

** Installation

Create and activate a virtual environment, for example with [[https://docs.conda.io/en/latest/][conda]]:

#+begin_src shell
  conda create -n spe python=3.10
  conda activate spe
#+end_src

Clone this repository:
#+begin_src shell
  git clone https://github.com/idiap/idiap_spe.git
#+end_src

Install the =idiap_spe= package in [[https://pip.pypa.io/en/stable/topics/local-project-installs/#editable-installs][editable mode]], so that you are free to make
any changes to the code:

#+begin_src shell
  cd idiap_spe
  pip install -e .
#+end_src

** Exercises

The following exercises are available as both [[https://jupyter.org/][Jupyter]] and [[https://orgmode.org/worg/org-contrib/babel/][Org Babel]] notebooks.

- Speech signal analysis ([[./jupyter-notebooks/speech_analysis.ipynb][Jupyter]], [[./org-notebooks/speech_analysis.org][Org]])
- Hidden Markov models ([[./jupyter-notebooks/hmm.ipynb][Jupyter]], [[./org-notebooks/hmm.org][Org]])
- Grapheme-to-phoneme (G2P) conversion ([[./jupyter-notebooks/g2p.ipynb][Jupyter]], [[./org-notebooks/g2p.org][Org]])

A Jupyter notebook can be launched like this:

#+begin_src shell
  conda activate spe
  jupyter notebook jupyter-notebooks/hmm.ipynb
#+end_src

*Note:* For the HMM exercise you might first need to install /Graphviz/
following the instructions for your operating system here:
[[https://graphviz.org/download/]]

** Acknowledgements

The speech signal analysis and HMM exercises were originally developed in Matlab
by Sacha Krstulović, Hervé Bourlard, and Mathew Magimai-Doss for the /Speech
Processing and Speech Recognition/ course at École polytechnique fédérale de
Lausanne (EPFL).


CHANGELOG.org
# SPDX-FileCopyrightText: Idiap Research Institute
# SPDX-FileContributor: Enno Hermann <enno.hermann@idiap.ch>
# SPDX-License-Identifier: GPL-3.0-only

* Changelog

All notable changes to this project will be documented in this file.

The format is based on [[https://keepachangelog.com/en/1.0.0/][Keep a Changelog]], and this project adheres to [[https://semver.org/spec/v2.0.0.html][Semantic
Versioning]].

** [Unreleased]

** [1.0.0] - 2023-11-07

*** Added

- =idiap_spe= Python package
- Speech signal analysis exercise
- Hidden Markov models exercise
- Grapheme-to-phoneme (G2P) exercise


tests/__init__.py
# SPDX-FileCopyrightText: Idiap Research Institute
#
# SPDX-License-Identifier: CC0-1.0

"""Unit tests for idiap_spe package."""


tests/test_g2p.py
# SPDX-FileCopyrightText: Idiap Research Institute
# SPDX-FileContributor: Enno Hermann <enno.hermann@idiap.ch>
#
# SPDX-License-Identifier: GPL-3.0-only

"""Unit tests for G2P exercise."""

import cmudict
import phonetisaurus as ps


def test_load_cmudict() -> None:
    """Test that the CMU lexicon can be correctly loaded."""
    lexicon = ps.load_lexicon(cmudict.dict_string().decode().split("\n"))
    assert lexicon["either"] == [["IY1", "DH", "ER0"], ["AY1", "DH", "ER0"]]


tests/test_hmm.py
# SPDX-FileCopyrightText: Idiap Research Institute
# SPDX-FileContributor: Enno Hermann <enno.hermann@idiap.ch>
#
# SPDX-License-Identifier: GPL-3.0-only

"""Unit tests for hidden Markov model module."""

import pytest

from idiap_spe.hmm import HMM, Gaussian


@pytest.fixture()
def gaussian() -> Gaussian:
    """Gaussian fixture."""
    return Gaussian(mean=[0, 0], cov=[[0, 0], [0, 0]], allow_singular=True)


def test_gmm_repr(gaussian: Gaussian) -> None:
    """Test how Gaussians are printed."""
    assert str(gaussian) == "mean=[0.0, 0.0], cov=[[0.0, 0.0], [0.0, 0.0]]"


def test_hmm_repr(gaussian: Gaussian) -> None:
    """Test how HMMs are printed."""
    hmm = HMM(labels=["x"], transitions=[[1.0]], gaussians=[gaussian])
    assert str(hmm) == (
        "HMM(transitions=[[1.0]], "
        "gaussians=[mean=[0.0, 0.0], cov=[[0.0, 0.0], [0.0, 0.0]]])"
    )


idiap_spe/hmm.py
# SPDX-FileCopyrightText: Idiap Research Institute
# SPDX-FileContributor: Enno Hermann <enno.hermann@idiap.ch>
#
# SPDX-License-Identifier: GPL-3.0-only

"""Hidden Markov models."""

from collections.abc import Sequence
from typing import TypedDict

import matplotlib as mpl
import matplotlib.pyplot as plt
import numpy as np
import numpy.typing as npt
import scipy.linalg
from cycler import cycler
from graphviz import Digraph
from scipy.stats._multivariate import multivariate_normal_frozen


class Gaussian(multivariate_normal_frozen):  # type: ignore[misc]
    """Multivariate Gaussian distribution."""

    def __repr__(self) -> str:
        """Represent a Gaussian by its means and covariances."""
        return f"mean={self.mean.tolist()}, cov={self.cov.tolist()}"


class HMM:
    """Hidden Markov Model for multivariate Gaussian-distributed observations.

    Keyword arguments:
    transitions -- Transition probability matrix
    gaussians -- List of Gaussians that define the emission probabilities
    labels -- Optional list of state labels
    """

    def __init__(
        self,
        *,
        transitions: npt.ArrayLike,
        gaussians: Sequence[Gaussian],
        labels: Sequence[str] | None = None,
    ) -> None:
        """Create an HMM."""
        self.transitions = np.atleast_1d(transitions)
        if len(self.transitions) != len(gaussians) or len(gaussians) == 0:
            msg = "Transition matrix must match number of Gaussians and be non-zero"
            raise ValueError(msg)

        # Transition probability matrix `A`
        with np.errstate(divide="ignore"):  # Ignore divide by zero warning
            self.log_transitions = np.log(self.transitions)

        # Emission probabilities `b(x_t)` modelled by Gaussians
        self.gaussians = gaussians

        # Optional state labels
        self.labels = labels if labels else len(gaussians) * [""]

        self.n_states = len(gaussians)
        self.dim = len(gaussians[0].mean)

    def __repr__(self) -> str:
        """Represent an HMM by its transitions and Gaussians."""
        return (
            f"HMM(transitions={self.transitions.tolist()}, gaussians={self.gaussians})"
        )

    def sample(
        self, *, plot: bool = False
    ) -> tuple[npt.NDArray[np.float64], list[int], list[str]]:
        """Draw a sequence of samples from the HMM.

        Keyword arguments:
        plot -- Whether to plot the sequence

        Returns:
        - Observations: (2, N - 2) array
        - States: (N, 1) array
        - List of state labels
        """

        def _sample_state(local_transitions: npt.NDArray[np.float64]) -> int:
            """Sample a state given local transition probabilities."""
            cumsum = np.cumsum(local_transitions)
            rng = np.random.default_rng()
            uniform = rng.random()
            state = 0
            while uniform >= cumsum[state]:
                state += 1
            return state

        # Generate the state sequence
        states = [0]  # Begin with the initial state
        labels = [self.labels[0]]
        t = 1
        while states[-1] != self.n_states - 1:
            t += 1
            state = _sample_state(self.transitions[states[-1]])
            states.append(state)
            labels.append(self.labels[state])

        # Generate observations for each state except initial and final
        observations = np.zeros((len(states) - 2, 2))
        for i, state in enumerate(states[1:-1]):
            observations[i] = self.gaussians[state].rvs()

        if plot:
            self.plot_sample(observations, states)
        return observations, states, labels

    def forward(self, observations: npt.NDArray[np.float64]) -> float:
        """Return log likelihood of the observed sequence w.r.t. the HMM.

        Uses the forward algorithm.
        """
        # Precompute the emission probabilities b(x), i.e. pdfs, for all
        # observations and emitting states
        log_bs = np.zeros((len(observations), self.n_states))
        for state in range(1, self.n_states - 1):
            log_bs[:, state] = np.log(self.gaussians[state].pdf(observations))

        # Compute the initial alphas
        alphas = np.ones((len(observations), self.n_states)) * -np.inf
        alphas[0] = self.log_transitions[0] + log_bs[0]

        # Forward recursion
        for t in range(1, len(observations)):
            for j in range(1, self.n_states - 1):
                alphas[t, j] = log_bs[t, j] + np.logaddexp.reduce(
                    alphas[t - 1] + self.log_transitions[:, j]
                )

        # Termination
        return float(
            np.logaddexp.reduce(alphas[-1] + self.log_transitions[:, self.n_states - 1])
        )

    def viterbi(
        self, observations: npt.NDArray[np.float64]
    ) -> tuple[npt.NDArray[np.float64], float]:
        """Return most likely state sequence and its log likelihood w.r.t. to the HMM.

        Uses the Viterbi algorithm.
        """
        # Precompute the emission probabilities b(x), i.e. pdfs, for all
        # observations and emitting states
        log_bs = np.zeros((len(observations), self.n_states))
        for state in range(1, self.n_states - 1):
            log_bs[:, state] = np.log(self.gaussians[state].pdf(observations))

        # Compute the initial deltas
        deltas = np.ones((len(observations), self.n_states)) * -np.inf
        deltas[0] = self.log_transitions[0] + log_bs[0]

        # Initialize backpointers
        pointers = np.zeros((len(observations), self.n_states), dtype=int)

        # Viterbi recursion
        for t in range(1, len(observations)):
            for j in range(1, self.n_states - 1):
                deltas[t, j] = (
                    np.max(deltas[t - 1] + self.log_transitions[:, j]) + log_bs[t, j]
                )
                pointers[t, j] = np.argmax(deltas[t - 1] + self.log_transitions[:, j])

        # Termination
        log_likelihood = np.max(deltas[-1] + self.log_transitions[:, self.n_states - 1])

        # Backtracking
        path = np.zeros((len(observations) + 2), dtype=int)
        path[-1] = self.n_states - 1
        path[-2] = np.argmax(deltas[-1] + self.log_transitions[:, self.n_states - 1])
        for t in range(len(observations) - 1, 0, -1):
            path[t] = pointers[t, path[t + 1]]
        path[0] = 0

        return path, log_likelihood

    def plot(self) -> Digraph:
        """Plot the states and transitions of the HMM."""
        g = Digraph()
        g.attr(rankdir="LR")  # left-to-right layout
        g.attr("edge", fontsize="10")

        # Initial and final state
        g.attr("node", shape="doublecircle")
        g.node("0", self.labels[0])
        g.node(str(len(self.transitions) - 1), self.labels[-1])

        # Other states
        g.attr("node", shape="circle")
        for i, label in enumerate(self.labels):
            g.node(str(i), label)

        # Edges
        for i in range(len(self.transitions)):
            for j in range(len(self.transitions)):
                if self.transitions[i][j] > 0:
                    g.edge(str(i), str(j), label=str(self.transitions[i][j]))
        return g

    def pprint(self) -> Digraph:
        """Print and plot a nice representation of the HMM."""
        print(f"States: {self.labels}\n")

        print("Transition matrix:")
        print(self.transitions)

        print("\nGraph:", end="")
        return self.plot()

    def plot_sample(
        self,
        observations: npt.NDArray[np.float64] | None = None,
        states: Sequence[int] | None = None,
    ) -> None:
        """Plot a sample of the HMM."""
        if observations is None:
            observations, states, _ = self.sample()

        _fig = plt.figure(figsize=(16, 8))
        ax0 = plt.subplot(2, 4, (1, 2))
        ax1 = plt.subplot(2, 4, (5, 6))
        ax2 = plt.subplot(1, 4, (3, 4))
        plt.subplots_adjust(wspace=0.5, hspace=0.3)
        gray = "0.8"
        _cmap = cycler(color=mpl.colormaps["Dark2"](range(self.n_states - 2)))

        ax0.set_title("First dimension")
        ax0.plot(observations[:, 0], c=gray)

        ax1.set_title("Second dimension")
        ax1.set_xlabel("Steps")
        ax1.plot(observations[:, 1], c=gray)

        ax2.set_title("2D")
        ax2.set_xlabel("First dimension")
        ax2.set_ylabel("Second dimension")
        ax2.plot(observations[:, 0], observations[:, 1], c=gray)

        if states is not None:
            # Discard the non-emitting initial and final states
            states_np = np.array(states)[1:-1, np.newaxis]
            if len(observations) != len(states_np):
                msg = "Number of observations and states does not match"
                raise ValueError(msg)

            for state in range(1, self.n_states - 1):
                filtered = np.where((states_np == state), observations, np.nan)
                ax0.plot(filtered[:, 0], marker="o", ls="")
                ax1.plot(filtered[:, 1], marker="o", ls="")
                (base,) = ax2.plot(
                    filtered[:, 0],
                    filtered[:, 1],
                    marker="o",
                    ls="",
                    label=f"State {state}: {self.labels[state]}",
                )
                ax2.plot(
                    self.gaussians[state].mean[0],
                    self.gaussians[state].mean[1],
                    marker="+",
                    mew=3,
                    ms=20,
                    color=base.get_color(),
                )
                t = np.linspace(-np.pi, np.pi, num=100)
                stdev = scipy.linalg.sqrtm(self.gaussians[state].cov)
                circle = np.array([np.cos(t), np.sin(t)]).T.dot(stdev) + np.tile(
                    self.gaussians[state].mean, (100, 1)
                )
                ax2.plot(circle[:, 0], circle[:, 1], color=base.get_color())

            ax2.legend()

        plt.show()  # type: ignore[no-untyped-call,unused-ignore]

    def compare_sequences(
        self,
        observations: npt.NDArray[np.float64],
        states1: npt.NDArray[np.float64],
        states2: npt.NDArray[np.float64],
    ) -> None:
        """Compare two alignments of the observation sequence."""
        plt.figure(figsize=(16, 8))
        ax0 = plt.subplot(2, 1, 1)
        ax1 = plt.subplot(2, 1, 2)
        plt.subplots_adjust(wspace=0.5, hspace=0.3)
        gray = "0.8"
        cycler(color=mpl.colormaps["Dark2"](range(self.n_states - 2)))

        ax0.set_title("First state sequence")
        ax0.plot(observations[:, 0], c=gray)

        ax1.set_title("Second state sequence")
        ax1.set_xlabel("Steps")
        ax1.plot(observations[:, 0], c=gray)

        # Discard the non-emitting initial and final states
        states1 = states1[1:-1, np.newaxis]
        states2 = states2[1:-1, np.newaxis]
        if not len(observations) == len(states1) == len(states2):
            msg = "Number of observations and states does not match"
            raise ValueError(msg)

        # Identify misalignments
        misalignments = np.where((states1 != states2), observations, np.nan)

        for state in range(1, self.n_states - 1):
            filtered1 = np.where((states1 == state), observations, np.nan)
            filtered2 = np.where((states2 == state), observations, np.nan)
            ax0.plot(
                filtered1[:, 0],
                marker="o",
                ls="",
                label=f"State {state}: {self.labels[state]}",
            )
            ax1.plot(filtered2[:, 0], marker="o", ls="")

        # Plot misalignments
        class PlotDict(TypedDict):
            marker: str
            ls: str
            markerfacecolor: str
            markeredgecolor: str
            markersize: int

        kwargs: PlotDict = {
            "marker": "o",
            "ls": "",
            "markerfacecolor": "none",
            "markeredgecolor": "r",
            "markersize": 15,
        }
        label = "No misalignments" if np.isnan(misalignments).all() else "Misalignments"
        ax0.plot(misalignments[:, 0], label=label, **kwargs)
        ax1.plot(misalignments[:, 0], **kwargs)

        ax0.legend()

        plt.show()  # type: ignore[no-untyped-call,unused-ignore]


idiap_spe/speech_analysis.py
# SPDX-FileCopyrightText: Idiap Research Institute
# SPDX-FileContributor: Enno Hermann <enno.hermann@idiap.ch>
#
# SPDX-License-Identifier: GPL-3.0-only

"""Functions for speech analysis."""

import importlib.resources
import math
from collections.abc import Callable

import librosa
import matplotlib.pyplot as plt
import numpy as np
import numpy.typing as npt
from matplotlib.axes import Axes
from scipy.signal import lfilter

plt.rcParams["figure.facecolor"] = "w"

SAMPLING_FREQUENCY = 16000

FIG_WIDTH = 15
FIG_HEIGHT = 5


def load_signal(filename: str) -> npt.NDArray[np.float64]:
    """Load a speech signal from the given file."""
    included_fn = importlib.resources.files("idiap_spe").joinpath(
        f"data/speech_analysis/{filename}"
    )
    if included_fn.is_file():
        filename = str(included_fn)
    with open(filename) as f:
        lines = [int(line.rstrip()) for line in f]
        return np.array(lines[16:], dtype="int64")  # Actual data starts at line 17


def _plot_grid(ax: Axes) -> None:
    """Add a grid to the plot."""
    ax.minorticks_on()
    ax.grid()
    ax.grid(which="minor", ls=":")


def plot_amplitude(
    signal: npt.NDArray[np.float64], ax: Axes, title: str = "", *, grid: bool = False
) -> None:
    """Plot amplitude of a signal."""
    if grid:
        _plot_grid(ax)
    ax.set_title(title)
    ax.set_xlabel("Time in samples")
    ax.set_ylabel("Amplitude")
    ax.set_xlim(0, len(signal))
    ax.set_ylim(-32768, 32767)
    ax.plot(signal, lw=1)


def plot_frequency(
    frequencies: npt.ArrayLike,
    amplitudes: npt.NDArray[np.float64],
    ax: Axes,
    title: str = "",
) -> None:
    """Plot frequency of a signal."""
    _plot_grid(ax)
    ax.set_title(title)
    ax.set_xlabel("Frequency (kHz)")
    ax.set_ylabel("Log amplitude")
    ax.set_xlim(0, max(np.atleast_1d(frequencies)))
    ax.plot(frequencies, amplitudes, lw=1)


def speech_signal_observation(
    filename: str, title: str = "Speech signal"
) -> npt.NDArray[np.float64]:
    """Load a speech signal from the given file and plot amplitude and log energy."""
    data = load_signal(filename)
    _, axs = plt.subplots(2, figsize=(FIG_WIDTH, FIG_HEIGHT * 2))
    plt.subplots_adjust(hspace=0.3)

    plot_amplitude(data, axs[0], title)

    # Plot energy
    frame_shift = 64
    frame_size = 256
    max_frames = math.floor((len(data) - frame_size + frame_shift) / frame_shift)
    energy = np.zeros(max_frames)

    for i in range(max_frames):
        loc = i * frame_shift
        window = data[loc : loc + frame_size]
        energy[i] = np.dot(window.T, window)

    log_energy = 10 * np.log10(energy)
    log_energy_norm = log_energy - np.mean(log_energy)

    axs[1].set_title("Short-time energy plot of the above utterance")
    axs[1].set_xlabel("Time in terms of frames")
    axs[1].set_ylabel("Log energy")
    axs[1].set_xlim(0, len(log_energy_norm))
    axs[1].set_ylim(0, max(log_energy_norm))
    axs[1].plot(log_energy_norm, ".")

    return data


def select_speech(
    data: npt.NDArray[np.float64],
    begin: int,
    end: int,
    title: str = "Windowed signal",
    *,
    plot: bool = True,
) -> npt.NDArray[np.float64]:
    """Return a window of the given speech signal and plot it.

    The data between frame numbers `begin` (inclusive) and `end` (exclusive) is
    returned.
    """
    window = data[begin:end]
    if plot:
        fig = plt.figure(figsize=(FIG_WIDTH, FIG_HEIGHT))
        ax = fig.gca()
        plot_amplitude(window, ax, title, grid=True)
    return window


def autocorrelation(
    data: npt.NDArray[np.float64],
    max_lags: int,
    title: str = "Autocorrelation",
    *,
    plot: bool = True,
) -> npt.NDArray[np.float64]:
    """Return and plot the autocorrelation of the given signal."""
    # Compute autocorrelation
    n = len(data)
    correlation = np.correlate(data, data, mode="full")
    correlation = correlation[n - 1 - max_lags : n + max_lags]

    # Plot autocorrelation
    if plot:
        fig, axs = plt.subplots(2, figsize=(FIG_WIDTH, FIG_HEIGHT * 2))
        plt.subplots_adjust(hspace=0.3)
        for i, x in enumerate([correlation, correlation[max_lags + 1 :]]):
            axs[i].minorticks_on()
            axs[i].grid()
            axs[i].grid(which="minor", ls=":")
            axs[i].set_xlabel("Lag number")
            axs[i].set_ylabel("Amplitude")
            axs[i].set_xlim((0, len(x)))
            axs[i].plot(x)

        axs[0].set_title(title)
        axs[1].set_title("Autocorrelation (right half)")

    return correlation


def _compute_log_fourier(
    data: npt.NDArray[np.float64], order: int = 512
) -> npt.NDArray[np.float64]:
    """Compute the log Fourier spectrum."""
    log_fourier: npt.NDArray[np.float64] = 20 * np.log10(
        np.abs(np.fft.fft(data, n=order))
    )
    return log_fourier


def fourier_spectrum(
    data: npt.NDArray[np.float64],
    order: int = 512,
    sf: int = SAMPLING_FREQUENCY,
    title: str = "Fourier spectrum",
) -> None:
    """Compute and plot the fourier spectrum of the given signal."""
    # Compute fourier spectrum
    log_fourier = _compute_log_fourier(data, order)
    frequencies = [n * sf / (order * 1000) for n in range(order)]

    # Plot fourier spectrum
    _fig, axs = plt.subplots(2, figsize=(FIG_WIDTH, FIG_HEIGHT * 2))
    plt.subplots_adjust(hspace=0.3)
    for i, (freq, amplitudes, plot_title) in enumerate(
        [
            (frequencies, log_fourier, title),
            (
                frequencies[: order // 2],
                log_fourier[: order // 2],
                "Fourier spectrum (left half)",
            ),
        ]
    ):
        plot_frequency(freq, amplitudes, axs[i], plot_title)


def spectrogram(
    data: npt.NDArray[np.float64],
    order: int,
    window: Callable[[int], npt.NDArray[np.float64]] = np.hanning,
    sf: int = SAMPLING_FREQUENCY,
    title: str = "Spectrogram",
) -> None:
    """Compute and plot the spectrogram of the given signal."""

    def _window_fun(x: npt.ArrayLike) -> npt.ArrayLike:
        a: npt.NDArray[np.float64] = np.array(x)
        return window(len(a)) * a

    plt.figure(figsize=(FIG_WIDTH, FIG_HEIGHT))
    plt.specgram(data, NFFT=order, window=_window_fun, Fs=sf, cmap="magma")
    plt.title(title)
    plt.xlabel("Time (s)")
    plt.ylabel("Frequency (Hz)")


def preemphasize(data: npt.NDArray[np.float64]) -> npt.NDArray[np.float64]:
    """Pre-emphasize to remove the spectral tilt due to glottal pulse spectrum."""
    diff_data = np.empty_like(data)
    diff_data[0] = data[0]
    diff_data[1:] = np.diff(data)
    return diff_data


def lp_spectrum(
    data: npt.NDArray[np.float64],
    lp_order: int,
    order: int,
    window: Callable[[int], npt.NDArray[np.float64]] = np.hanning,
    *,
    sf: int = SAMPLING_FREQUENCY,
    plot: bool = True,
) -> npt.NDArray[np.float64]:
    """Compute and plot the LP spectrum."""
    windowed_data = preemphasize(data)
    windowed_data = window(len(windowed_data)) * windowed_data

    frequencies = [n * sf / (order * 1000) for n in range(order // 2)]

    # Computation of LP and Fourier spectrum
    coefficients = librosa.lpc(windowed_data, order=lp_order)
    lp_spec: npt.NDArray[np.float64] = -20.0 * np.log10(
        np.abs(np.fft.fft(coefficients, order))
    )
    log_fourier = _compute_log_fourier(data, order)

    if plot:
        _fig, axs = plt.subplots(2, figsize=(FIG_WIDTH, FIG_HEIGHT * 2))
        plt.subplots_adjust(hspace=0.3)
        for i, (spectrum, title) in enumerate(
            [(log_fourier, "Fourier spectrum"), (lp_spec, "LP spectrum")]
        ):
            plot_frequency(frequencies, spectrum[: order // 2], axs[i], title)
    return lp_spec


def lp_residual(
    data: npt.NDArray[np.float64],
    lp_order: int,
    window: Callable[[int], npt.NDArray[np.float64]] = np.hanning,
    *,
    plot: bool = True,
) -> npt.NDArray[np.float64]:
    """Compute and plot the LP residual."""
    # Pre-emphasize, window the signal and compute LPC
    windowed_data = preemphasize(data)
    windowed_data = window(len(windowed_data)) * windowed_data
    coefficients = librosa.lpc(windowed_data, order=lp_order)

    residual = np.empty_like(data)
    padded_data = np.pad(data, (lp_order, 0))

    for i in range(len(data)):
        predict = 0
        for j in range(1, lp_order + 1):
            predict = predict + coefficients[j] * padded_data[i + lp_order - j]
        residual[i] = padded_data[i + lp_order] + predict

    # Plot signal and residual
    if plot:
        _fig, axs = plt.subplots(2, figsize=(FIG_WIDTH, FIG_HEIGHT * 2))
        plt.subplots_adjust(hspace=0.3)
        for i, (signal, title) in enumerate(
            [(data, "Original signal"), (residual, "LP residual signal")]
        ):
            plot_amplitude(signal, axs[i], title, grid=True)
    return residual


def speaker_variation(
    utterances: list[tuple[str, int]],
    lp_order: int = 16,
    sample_length: int = 480,
    fft_order: int = 512,
    sf: int = SAMPLING_FREQUENCY,
) -> None:
    """Plot the LP spectra for a list of utterances."""
    frequencies = [n * sf / (fft_order * 1000) for n in range(fft_order // 2)]
    fig = plt.figure(figsize=(FIG_WIDTH, FIG_HEIGHT))
    ax = fig.gca()
    ax.set_xlabel("Frequency (kHz)")
    ax.set_ylabel("Log amplitude")
    ax.set_title("Comparison of LP spectra")
    ax.set_xlim(0, sf // 2000)

    for filename, start in utterances:
        st_data = load_signal(filename)[start : start + sample_length]
        spectrum = lp_spectrum(st_data, lp_order, fft_order, plot=False)
        ax.plot(frequencies, spectrum[: fft_order // 2], label=filename)
    ax.legend()


def sift(
    filename: str,
    lp_order: int = 10,
    frame_size: int = 480,
    frame_shift: int = 160,
    sf: int = SAMPLING_FREQUENCY,
) -> list[float]:
    """Compute the pitch contour using the SIFT algorithm."""
    max_lags = 256

    b = [0.0357081667, -0.0069956244, -0.0069956244, 0.0357081667]
    a = [1.0, -2.34036589, 2.01190019, -0.61419218]

    data = load_signal(filename)
    max_frames = int(np.floor((len(data) - frame_size + frame_shift) / frame_shift))
    divisor = 1.8

    pitch = []
    for i in range(max_frames):
        idx = i * frame_shift
        frame = data[idx : idx + frame_size]
        frame = lfilter(b, a, frame)
        residual = lp_residual(frame, lp_order, plot=False)
        residual = autocorrelation(residual, max_lags, plot=False)

        max_residual = residual[max_lags] / divisor
        max_idx = max_lags
        for j in range(max_lags + (sf // 400), max_lags + (sf // 80)):
            if residual[j] > max_residual:
                max_residual = residual[j]
                max_idx = j

        max_idx = max_idx - max_lags
        if max_idx > 0:
            pitch.append(sf / max_idx)
        else:
            pitch.append(0.0)

    _fig, axs = plt.subplots(2, figsize=(FIG_WIDTH, FIG_HEIGHT * 2))
    plt.subplots_adjust(hspace=0.3)

    # Plot signal
    plot_amplitude(data, axs[0], "Speech signal")

    # Plot pitch
    axs[1].set_title("Pitch contour")
    axs[1].set_xlabel("Time in samples")
    axs[1].set_ylabel("Pitch frequency (Hz)")
    axs[1].set_xlim(0, len(pitch))
    axs[1].plot(pitch, ".")

    return pitch
